### PR TITLE
Vim movement multipliers

### PR DESCRIPTION
**Pull request by ryabouza**
```
        _,'|             _.-''``-...___..--';)
       /_ \'.      __..-' ,      ,--...--'''
      <\    .`--'''       `     /'
       `-';'               ;   ; ;
 __...--''     ___...--_..'  .;.'
(,__....----'''       (,..--''   
```

### CODE REVIEW

File: ce_config.c [hunk 1]
```diff
@@ -377,13 +380,127 @@ void enter_insert_mode(ConfigState* config_state, Point* cursor)
...
-     if(config_state->command_key == '\0'){
-          config_state->command_key = key;
-          return false;
+     config_state->command_multiplier = 0;
+     config_state->command_key = '\0';
+     config_state->movement_multiplier = 0;
+     memset(config_state->movement_keys, 0, sizeof config_state->movement_keys);
+}
+
+// returns true if key may be interpretted as a multiplier given the current mulitplier state
```
Comment (robber-m) [inline ce_config.c]: interpreted*

File: ce_config.c [hunk 2]
```diff
@@ -377,13 +380,127 @@ void enter_insert_mode(ConfigState* config_state, Point* cursor)
...
+bool key_is_multiplier(uint64_t multiplier, char key)
+{
+     return (key >='1' && key <= '9') || (key == '0' && multiplier > 0);
```
Comment (robber-m) [inline ce_config.c:434]: Movements cannot start with a 0 correct?

Comment (ryabouza) [reply ce_config.c:434]: Yes. '0' is a command (move to beginning of line).

**Comment by ryabouza** [comment]
We really need unit tests. I definitely broke everything.

File: ce.c [hunk 3]
```diff
@@ -377,8 +377,12 @@ char* ce_dupe_string(Buffer* buffer, const Point* start, const Point* end)
      CE_CHECK_PTR_ARG(start);
      CE_CHECK_PTR_ARG(end);
 
-     if(!ce_point_on_buffer(buffer, start)) return NULL;
-     if(!ce_point_on_buffer(buffer, end)) return NULL;
+     if(!ce_point_on_buffer(buffer, start) && start->x != (int64_t)strlen(buffer->lines[start->y])){
```
Comment (justy989) [inline ce.c:380]: Hmmm, we seem to have this problem in multiple places where we sometimes we want to check if the point is on the buffer but less than the end of the line, and sometimes we are ok with it being at the end of the line. Maybe we want 2 functions, something like: ce_point_in_buffer() vs ce_point_on_buffer() Something to think about in the future, not necessary to fix for this pull request.

File: ce_config.c [hunk 4]
```diff
@@ -377,13 +420,265 @@ void enter_insert_mode(ConfigState* config_state, Point* cursor)
...
+     config_state->command_multiplier = 0;
+     config_state->command_key = '\0';
+     config_state->movement_multiplier = 0;
+     memset(config_state->movement_keys, 0, sizeof config_state->movement_keys);
+}
+
+// returns true if key may be interpreted as a multiplier given the current mulitplier state
+bool key_is_multiplier(uint64_t multiplier, char key)
+{
+     return (key >='1' && key <= '9') || (key == '0' && multiplier > 0);
```
Comment (justy989) [inline ce_config.c:434]: dang '0' is tricky !

File: ce.c (hunk 3 above)
Comment (ryabouza) [reply ce.c:380]: Hahahaha, I came up with the same function names when talking to @bobbymcshane

**Review by justy989 — APPROVED** [review]

**Pull request merged**